Fuser/runner_util.py

```diff
@@ -0,0 +1,128 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Utility functions for the compose worker."""
+
+import multiprocessing
+import os
+import signal
+import sys
+import time
+from pathlib import Path
+
+
+def _kill_process_group(pid: int) -> None:
+    """Kill the process group of pid. Similar behavior as os.killpg."""
+    try:
+        os.killpg(pid, signal.SIGTERM)
+    except (ProcessLookupError, PermissionError):
+        pass
+    # Give processes a moment to terminate gracefully
+    time.sleep(0.1)
+    try:
+        os.killpg(pid, signal.SIGKILL)
+    except (ProcessLookupError, PermissionError):
+        pass
+
+
+def _run_candidate_process(
+    exec_filename: str,
+    run_dir: Path,
+    argv: list[str],
+    env: dict[str, str],
+    stdout_path: Path,
+    stderr_path: Path,
+) -> None:
+    """Target function for multiprocessing.Process that executes the candidate."""
+    # Create a new process group to contain potential subprocesses for cancel/timeout.
+    # This is similar to subprocess.Popen(start_new_session=True) behavior.
+    os.setpgrp()
+
+    with stdout_path.open("wb") as f_out, stderr_path.open("wb") as f_err:
+        os.chdir(str(run_dir))
+        # Replace the child's environment (similar to subprocess.Popen(...env=env)).
+        # This sets up the environment for the candidate execution.
+        os.environ.clear()
+        os.environ.update(env)
+        # Redirect stdout/stderr
+        sys.stdout.flush()
+        sys.stderr.flush()
+        os.dup2(f_out.fileno(), sys.stdout.fileno())
+        os.dup2(f_err.fileno(), sys.stderr.fileno())
```

Comment:
Prefer duplicating directly to fd 1 and 2 (stdout/stderr) instead of relying on `sys.stdout.fileno()` / `sys.stderr.fileno()`